phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexUpgradeTool.java

```diff
@@ -318,53 +323,155 @@ public int executeTool() {
-                return -1;
             }
+        } catch (IOException | SQLException e) {
+            LOGGER.severe("Something went wrong while executing "
+                    + operation + " steps for "+ dataTableFullNameGlobal + " " + e);
+            handleFailure(queryServices, dataTableFullNameGlobal, mutableTables);
         }
         // Opportunistically kick-off index rebuilds after upgrade operation
-        if (upgrade) {
-            for (String dataTableFullName : tablesAndIndexes.keySet()) {
-                rebuildIndexes(conn, conf, dataTableFullName);
-                LOGGER.info("Started index rebuild post " + operation + " of "
-                        + dataTableFullName);
+        rebuildIndexes(conn, conf, mutableTables);
+    }
+
+    private void enableImmutableTables(ConnectionQueryServices queryServices,
+            ArrayList<String> immutableList,
+            long startWaitTime) {
+        int retryCount = 0;
+        long endWaitTime = System.currentTimeMillis();
+        long waitMore = getWaitMoreTime(endWaitTime, startWaitTime);
+        while  (retryCount < 5 && !waited) {
+            // If the table is immutable, we need to wait for clients to purge
+            // their caches of table metadata
+
+            LOGGER.info("waiting for more " + waitMore + " ms for client cache "
+                    + "to expire for immutable tables");
+            try {
+                startWaitTime = System.currentTimeMillis();
+                Thread.sleep(waitMore);
+                waited = true;
+            } catch (InterruptedException e) {
+                endWaitTime = System.currentTimeMillis();
+                waitMore = getWaitMoreTime(endWaitTime, startWaitTime);
+                retryCount++;
+                LOGGER.warning("Sleep before starting index rebuild is interrupted. "
+                        + "Attempting to sleep again! " + e.getMessage());
             }
         }
-        return 0;
+        if (retryCount == 5) {
+            LOGGER.info("Sleep was interrupted multiple times for immutable tables, "
+                    + "we'll still go ahead and enable the table.");
+        }
+        try (Admin admin = queryServices.getAdmin()) {
+            for (String dataTableFullName: immutableList) {
+                dataTableFullNameGlobal = dataTableFullName;
+                HashSet<String> indexes = tablesAndIndexes.get(dataTableFullName);
+                enableTable(admin, dataTableFullName, indexes);
+            }
+        } catch (IOException | SQLException e) {
+            LOGGER.severe("Something went wrong while enabling immutable table "
+                    + dataTableFullNameGlobal);
+            //removing to avoid any rebuilds after upgrade
+            tablesAndIndexes.remove(dataTableFullNameGlobal);
+            immutableList.remove(dataTableFullNameGlobal);
+            throw new RuntimeException("Manually enable the table " + dataTableFullNameGlobal
```

Comment:
If enabling table 2 of 5 fails, you need to manually enable tables 2,3,4, and 5, and the error message should say so.